### PR TITLE
Fix conflicting routes for multiple ENIs in IPAM mode

### PR DESCRIPTION
**Pull request by recollir**
Discovered during investigation of #18868

See comment https://github.com/cilium/cilium/issues/18868#issuecomment-1077682444 of #18868 for more details on the conflicting routes for health endpoints that are left behind during restarts.

In particular we see that the health endpoint gets 2 routes associated with it. When the pod restarts it leaves one route behind, which results in a routing conflict and AWS dropping the packet when the IP is then reused for a real workload pod.

Our idea in solving this is to coalesce the 2 or more CIDR ranges into one.

Cheers,
@dezmodue
@recollir

Signed-off-by: Simone Sciarrati <s.sciarrati@gmail.com>
Signed-off-by: Federico Hernandez <f@ederi.co>

### CODE REVIEW

**Review by jibi — CHANGES_REQUESTED** [review]
Thanks for the contribution @recollir!

Could you please:
* update the commit title and description with an explanation of the changes
* add the `Signed-off-by` tag
* get rid of the 2 merge commits

**Comment by recollir** [comment]
@jibi I have now edited the commit msg and included the sign-off tags as well as deleted the 2 merge commits.

Please let me know if there is anything else that we should do. We have tested the change locally and in a dev cluster, but are unsure on the _style_ and _place_ of the change. Any feedback is appreciated. Thanks.

File: daemon/cmd/ipam.go
Comment (gandro) [inline daemon/cmd/ipam.go]: This seems identical to the code block below? Can you extract it into a separate function? Thanks

File: pkg/datapath/loader/base.go
Comment (gandro) [inline pkg/datapath/loader/base.go]: Taking the `&cidr` address here looks like a bug: https://github.com/golang/go/wiki/CommonMistakes#using-reference-to-loop-iterator-variable

**Review by gandro — CHANGES_REQUESTED** [review]

File: daemon/cmd/ipam.go
Comment (recollir) [reply daemon/cmd/ipam.go]: Moved to coalesce functionality to a function.

File: pkg/datapath/loader/base.go
Comment (recollir) [reply pkg/datapath/loader/base.go]: Fixed bug according with suggestion from link. Thank you.

**Comment by recollir** [comment]
@gandro we pushed your suggested changes.

**Comment by recollir** [comment]
A simple test using the function gives

```
[10.105.0.0/16 10.0.0.0/8] => [10.0.0.0/8]
[10.105.0.0/16 192.168.1.0/24] => [10.105.0.0/16 192.168.1.0/24]
[10.105.0.0/16 192.168.1.0/24 10.0.0.0/8] => [10.0.0.0/8 192.168.1.0/24]
```

based on

```
  CIDR := []string{"10.105.0.0/16","10.0.0.0/8"}
  newCIDR := coalesceCIDRs(CIDR)
  fmt.Printf("%v => %v\n",CIDR,newCIDR)

  CIDR = []string{"10.105.0.0/16","192.168.1.0/24"}
  newCIDR = coalesceCIDRs(CIDR)
  fmt.Printf("%v => %v\n",CIDR,newCIDR)

  CIDR = []string{"10.105.0.0/16","192.168.1.0/24","10.0.0.0/8"}
  newCIDR = coalesceCIDRs(CIDR)
  fmt.Printf("%v => %v\n",CIDR,newCIDR)
```

**Review by gandro — COMMENTED** [review]
Awesome, thanks! One last request: Could you squash the two commits together?

**Comment by gandro** [comment]
> @gandro Do you want the unit test in the same single commit? Or do you want to get the squash now as a first contribution and the unit test in a second PR?

Let's do the unit test in the same commit. There are no hard rules here, but to make it easier to read the git history in master, I think a single commit containing everything is the preferred way.

File: daemon/cmd/ipam_test.go
Comment (gandro) [inline daemon/cmd/ipam_test.go:6]: Please add the copyright header as found in other files, and make sure to `gofmt` this file. Thanks

**Review by gandro — COMMENTED** [review]
Almost there, thanks! The last commit needs also needs to have the `signed-off` line in addition to the inline feeback

File: daemon/cmd/ipam_test.go
Comment (recollir) [reply daemon/cmd/ipam_test.go:6]: Done.

**Comment by recollir** [comment]
@gandro We should be good to go now.

- added a unit test.
- ensured gofmt on it.
- added copyright to unit test.
- squashed everything to one commit.
- ensured sign off by me and Simone.

Thanks for guiding us through this.

File: daemon/cmd/ipam_test.go
Comment (gandro) [inline daemon/cmd/ipam_test.go]:
```suggestion
//go:build !privileged_tests
```

This is not an integration test relying on e.g. the kvstore, so it doesn't need the tag.

**Review by gandro — COMMENTED** [review]
Awesome. One more thing, then this is good to go!

**Review by gandro — APPROVED** [review]
Awesome, thanks for taking care of this! :rocket:

**Comment by recollir** [comment]
@gandro There is a failing test due to a missing file. https://github.com/cilium/cilium/runs/6981297570?check_suite_focus=true#step:11:29

It seems to be an example documentation file. How do you want us to deal with this? Are we nevertheless fine with the PR or should we take some action?

**Comment by gandro** [comment]
@recollir Ah yes, you will have to rebase your branch on the current master. The test suite is running against tests present in `master`, but the code under test will be built from your branch. This can cause test failures if the branch under test is a bit older, which seems to be the case for you (lagging ~130 commits behind). 

Rebasing it on the current master should fix it. Sorry about the inconvenience

**Comment by recollir** [comment]
@jibi We worked with @gandro on the PR which we finalised yesterday. From what I see your requested changes are all implemented now. Is your approval still needed?

@gandro What is the ongoing procedure for the PR now that you have approved it? Is anything more needed from us for the merge? When are you doing that? Curious to learn on your process.

**Comment by gandro** [comment]
> @gandro What is the ongoing procedure for the PR now that you have approved it? Is anything more needed from us for the merge? When are you doing that? Curious to learn on your process.

Ideally we want approval from all affected code owners. So let's give @jibi some time to re-review, because I'm also not super familiar with all the affected parts here.

In parallel, we also want to make sure this passes our CI testing (thus my `/test` command triggering CI above). There have been a few failures. You can take a look if you want, but I will also triage this later, to see if they are real failures that need to be addressed or if they are unrelated (our CI is a bit flaky at the moment).

File: daemon/cmd/ipam.go
Comment (jibi) [inline daemon/cmd/ipam.go]: nit: I think we can make this a bit more readable with:

```suggestion
		ip, mask, _ := net.ParseCIDR(k)
		cidrs = append(cidrs, &net.IPNet{IP: ip, Mask: mask.Mask})
```

**Comment by recollir** [comment]
Fair enough @gandro . We will watch any comments/coming in. Though replies from us might be delayed starting next week due to summer. :smiley:

File: pkg/datapath/loader/base.go
Comment (jibi) [inline pkg/datapath/loader/base.go]:
```suggestion
	for i := range cidrs {
		cidrs2 = append(cidrs2, &cidrs[i])
	}
```

File: pkg/datapath/loader/base.go
Comment (jibi) [inline pkg/datapath/loader/base.go]:
```suggestion
	cidrs = make([]net.IPNet, 0, len(resultcidr))
	for _, cidr := range resultcidr {
		cidrs = append(cidrs, *cidr)
	}
```

**Comment by gandro** [comment]
@recollir The failures look legitimate. 

There seems to be a nil-pointer panic in `allocateDatapathIPs`.  `cilium-agent` logs from one of the Jenkins runs:

```
2022-06-21T14:50:07.584387758Z [signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x25e2ccd]
2022-06-21T14:50:07.584436558Z 
2022-06-21T14:50:07.584621780Z goroutine 1 [running]:
2022-06-21T14:50:07.585014421Z github.com/cilium/cilium/daemon/cmd.(*Daemon).allocateDatapathIPs(0xc000339180, {0x3424da0, 0x52aedc0})
2022-06-21T14:50:07.585285431Z 	/go/src/github.com/cilium/cilium/daemon/cmd/ipam.go:263 +0x3ed
2022-06-21T14:50:07.585760387Z github.com/cilium/cilium/daemon/cmd.(*Daemon).allocateRouterIPv6(0xc000cdee80?, {0x3424da0?, 0x52aedc0?})
2022-06-21T14:50:07.585996698Z 	/go/src/github.com/cilium/cilium/daemon/cmd/ipam.go:193 +0x185
2022-06-21T14:50:07.586157544Z github.com/cilium/cilium/daemon/cmd.(*Daemon).allocateIPs(0xc000339180)
2022-06-21T14:50:07.586410844Z 	/go/src/github.com/cilium/cilium/daemon/cmd/ipam.go:407 +0xe5
2022-06-21T14:50:07.587103380Z github.com/cilium/cilium/daemon/cmd.NewDaemon({0x34215d0?, 0xc0001518c0}, 0xc0000a7b40, 0xc00033cd20, {0x3437db8, 0xc000154180})
2022-06-21T14:50:07.587356540Z 	/go/src/github.com/cilium/cilium/daemon/cmd/daemon.go:1088 +0x67c5
2022-06-21T14:50:07.587470228Z github.com/cilium/cilium/daemon/cmd.runDaemon()
2022-06-21T14:50:07.587750052Z 	/go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:1713 +0x7da
2022-06-21T14:50:07.588316113Z github.com/cilium/cilium/daemon/cmd.glob..func1(0x4d27140?, {0x2ecccb4?, 0x1?, 0x1?})
2022-06-21T14:50:07.588558560Z 	/go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:128 +0x3aa
2022-06-21T14:50:07.589569177Z github.com/spf13/cobra.(*Command).execute(0x4d27140, {0xc0001ac010, 0x1, 0x1})
2022-06-21T14:50:07.589574826Z 	/go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:860 +0x663
2022-06-21T14:50:07.589576997Z github.com/spf13/cobra.(*Command).ExecuteC(0x4d27140)
2022-06-21T14:50:07.589578897Z 	/go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:974 +0x3b4
2022-06-21T14:50:07.589580818Z github.com/spf13/cobra.(*Command).Execute(...)
2022-06-21T14:50:07.589582568Z 	/go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:902
2022-06-21T14:50:07.589584417Z github.com/cilium/cilium/daemon/cmd.Execute()
2022-06-21T14:50:07.589586308Z 	/go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:142 +0x45
2022-06-21T14:50:07.589588093Z main.main()
2022-06-21T14:50:07.589589842Z 	/go/src/github.com/cilium/cilium/daemon/main.go:15 +0x17
```

Travis CI also seems to be failing due to this panic: https://app.travis-ci.com/github/cilium/cilium/jobs/574223538

This needs to be addressed before we can merge this PR.

File: daemon/cmd/ipam.go
Comment (gandro) [inline daemon/cmd/ipam.go]: I believe the issue is that `result` can be nil if the `if routerIP != nil {` branch is never taken.

**Review by gandro — CHANGES_REQUESTED** [review]

File: daemon/cmd/ipam.go
Comment (recollir) [reply daemon/cmd/ipam.go]: Thank you for the hint and pointer (no pun intended) to this. We will try to address it as fast as we can.

**Review by jibi — COMMENTED** [review]
Few nits related to making the code a bit more readable

File: pkg/datapath/loader/base.go
Comment (dezmodue) [reply pkg/datapath/loader/base.go]: Changed as per suggestion, thank you

File: pkg/datapath/loader/base.go
Comment (dezmodue) [reply pkg/datapath/loader/base.go]: Changed as per suggestion, thank you

File: daemon/cmd/ipam.go
Comment (dezmodue) [reply daemon/cmd/ipam.go]: Changed as per suggestion, thank you

File: daemon/cmd/ipam.go
Comment (dezmodue) [reply daemon/cmd/ipam.go]: : D I wish I had seen this comment - I have just pushed a change to wrap the call to coalesceCIDRs in a check for nil. I added some debug statements and noticed the same. Hopefully the changes just pushed address the issue (local testing in Vagrant is OK).

**Comment by dezmodue** [comment]
Added tests for ipv6

**Comment by dezmodue** [comment]
The "[build-and-push-prs (cilium, ./images/cilium/Dockerfile)](https://github.com/cilium/cilium/runs/7029419803?check_suite_focus=true#logs)" job seems stuck on the "Load cilium Golang cache build from GitHub" step - it's been logging for a very long time:

```
Received` 838476060 of 842670364 (99.5%), 0.1 MBs/sec
Received 838476060 of 842670364 (99.5%), 0.1 MBs/sec
Received 838476060 of 842670364 (99.5%), 0.1 MBs/sec
Received 838476060 of 842670364 (99.5%), 0.1 MBs/sec
Received 838476060 of 842670364 (99.5%), 0.1 MBs/sec
```

**Comment by dezmodue** [comment]
Good morning @jibi @gandro, ready to be tested again.

**Comment by dezmodue** [comment]
Hi @gandro I looked into the error for the k8s-1.16-kernel-4.9 (test-1.16-4.9) but it is not clear to me what to do next, any guidance appreciated

**Comment by gandro** [comment]
> Hi @gandro I looked into the error for the k8s-1.16-kernel-4.9 (test-1.16-4.9) but it is not clear to me what to do next, any guidance appreciated

Hey @dezmodue - looking at the `cilium-agent` logs in the artifact file, I see the following timeout:
```
7622:2022-06-27T10:39:20.882218066Z level=error msg="Cannot create CEP" containerID= controller="sync-to-k8s-ciliumendpoint (579)" datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=579 error="etcdserver: request timed out" identity=25824 ipv4= ipv6= k8sPodName=/ subsys=endpointsynchronizer
```

This looks like a ephemeral issue, I'll restart the test.

**Comment by gandro** [comment]
/test-1.16-4.9

**Review by gandro — APPROVED** [review]

**Comment by recollir** [comment]
Thank you for restarting the test.

**Comment by recollir** [comment]
Seems that the tests pass now. @gandro Should we do a final rebase on master. We did not think about that this morning when requesting the new test round. We are about 23 commits behind.

**Comment by gandro** [comment]
> Seems that the tests pass now. @gandro Should we do a final rebase on master. We did not think about that this morning when requesting the new test round. We are about 23 commits behind.

If tests have passed, then there is no need to rebase. It's also useful to keep the current CI runs around, they will be lost if you rebase.

We usually only rebase if CI has been updated (which happened to you earlier), or if the branch is very old (e.g. older than say a week) to avoid hidden merge conflicts (e.g. where an API changed somewhere else in the code). I don't think this is the case here, so no need to rebase.

Therefore let's wait for the remaining reviews from @jibi and then this should be ready to merge :rocket:

**Review by jibi — APPROVED** [review]
LGTM! :rocket:

**Comment by recollir** [comment]
🍾 Thanks. @gandro and @jibi

**Comment by dezmodue** [comment]
Hi @gandro @jibi, will this fix be back ported to 1.10 and 1.11? We'd be interested since we're on 1.10 right now. Thanks

**Comment by gandro** [comment]
> Hi @gandro @jibi, will this fix be back ported to 1.10 and 1.11? We'd be interested since we're on 1.10 right now. Thanks

Thanks for reminding me. Yes, we want to backport this.

**Comment by recollir** [comment]
@gandro Hi there. Your comment arrived us in the middle of our summer vacation and us being AFK... Sorry for not leaving a short ping though. We are now back and will read through your comments and suggestions. We will probably have questions. I think working on this together would probably be a good thing.

**Comment by gandro** [comment]
@recollir Great to hear! Unfortunately I'll be out till August 22 now as well, but I'll happily take a look once I'm back. In the meantime, I'm sure other Cilium committers can also provide feedback based on my above write up.

**Comment by recollir** [comment]
@gandro No worries. Well deserved to be off and enjoy the summer. We have looked at your comments in the meantime, think we understand the changes needed and will start to work on those. 🌞

**Comment by recollir** [comment]
Question though: should we do a new PR or work in this one?

**Comment by dezmodue** [comment]
Hi @gandro the changes requested are addressed in https://github.com/cilium/cilium/pull/20848

**Pull request merged**
